### ISSUE TITLE
false-positive -Wreturn-type error

### ISSUE THREAD

**Issue by 7a12ef1e-274b-41a0-a5b2-ec732fce9bc9** [llvm/llvm-project#51815]
|  |  |
| --- | --- |
| Bugzilla Link | [52473](https://llvm.org/bz52473) |
| Version | trunk |
| OS | MacOS X |
| CC | @comanea,@DougGregor,@zygoloid |

## Extended Description 
This is a regression somewhere between clang revision 
`llvmorg-14-init-7378-gaee49255` and `llvmorg-14-init-8281-gaa1d32f5 - build 7378 has expected behavior. 

Repro: 
1.  Save code below to file file `a.m`
2.  Compile code: `bin/clang -Wreturn-type -Wno-objc-root-class -c -o a.o a.m`

```
@&#8203;interface Bar
- (Bar*)baz;
- (Bar*)quux;
- (int)query;
@&#8203;end
@&#8203;implementation Bar
- (Bar* ) baz
{
    @&#8203;try {
        @&#8203;throw self;
    } @&#8203;catch (Bar*) {
        return self;
    }
    return (Bar*)0;
}
@&#8203;end
```

Expected behavior: 
  compilation is free from warnings

Actual behavior:
  $ bin/clang -Wreturn-type -Wno-objc-root-class -c -o a.o a.m
  a.m:14:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
1 warning generated.

**Comment by nico** [comment]
Fixed in 8874ada906f6a2ae0c4aa978581a0c9c26b89ffd
